src/game/Grids/ObjectGridLoader.cpp

```diff
@@ -118,47 +118,61 @@ void LoadHelper(CellGuidSet const& guid_set, CellPair& cell, GridRefManager<T>&
 
     for (uint32 guid : guid_set)
     {
-        T* obj;
+        T* obj = nullptr;
         uint32 newGuid = guid;
         if constexpr (std::is_same_v<T, GameObject>)
         {
             GameObjectData const* data = sObjectMgr.GetGOData(guid);
             MANGOS_ASSERT(data);
-            obj = (T*)GameObject::CreateGameObject(data->id);
+            obj = static_cast<T*>(GameObject::CreateGameObject(data->id));
             if (map->GetSpawnManager().IsEventGuid(guid, HIGHGUID_GAMEOBJECT))
                 newGuid = 0;
         }
         else
         {
-            obj = new T;
+            obj = new T();
             if (map->GetSpawnManager().IsEventGuid(guid, HIGHGUID_UNIT))
                 newGuid = 0;
         }
         // sLog.outString("DEBUG: LoadHelper from table: %s for (guid: %u) Loading",table,guid);
-        if (!obj->LoadFromDB(guid, map, newGuid, 0))
+        if (!obj || !obj->LoadFromDB(guid, map, newGuid, 0))
         {
             delete obj;
             continue;
         }
+ 
+        // Set cell info (used also by creatures)
+        addUnitState(obj, cell);
 
+        // Only gameobjects are added to grid here
         if (!obj->IsCreature())
-        {
             grid.AddGridObject(obj);
+        
+        // Check whether object is correctly in world
+        if (!obj->IsInWorld())
```

Comment:
If this happens, we have a critical problem in the core and need to shut down immediately. The assert is there for a critical reason.